extensions/amp-form/0.1/form-verifiers.js

```diff
@@ -0,0 +1,418 @@
+export class AsyncVerifier extends FormVerifier {
+  /**
+   * @param {!HTMLFormElement} form
+   * @param {!Array<!VerificationGroup>} groups
+   * @param {function():Promise<!../../../src/service/xhr-impl.FetchResponse>} xhr
+   */
+  constructor(form, groups, xhr) {
+    super(form, groups);
+
+    /** @protected @const*/
+    this.doXhr_ = xhr;
+
+    /** @protected {?LastAddedResolver} */
+    this.xhrResolver_ = null;
+
+    /** @private {?Promise} */
+    this.xhrVerifyPromise_ = null;
+  }
+
+  /**
+   * Returns a promise that resolves when xhr verify finishes. the promise
+   * will be null if verify submit has not started.
+   * @visibleForTesting
+   */
+  xhrVerifyPromiseForTesting() {
+    return this.xhrVerifyPromise_;
+  }
+
+  /** @override */
+  onMutate(input) {
+    const group = this.getGroup_(input);
+    if (group) {
+      group.setDirty(true);
+      group.clearErrors();
+    }
+  }
+
+  /** @override */
+  onCommit(input, afterVerify) {
+    if (this.isVerificationElement_(input)) {
+      this.maybeVerify_(afterVerify);
+    }
+  }
+
+  /**
+   * Sends the verify request if any group is ready to verify.
+   * @param {!function(!Array<!Element>)} afterVerify
+   * @private
+   */
+  maybeVerify_(afterVerify) {
+    if (this.shouldVerify_()) {
+      const xhrConsumeErrors = this.doXhr_().then(() => {
+        return [];
+      }, error => {
+        // A 400 error response should be handled here.
+        return getResponseErrorData_(
+            /** @type {!../../../src/service/xhr-impl.FetchError} */ (error));
+      });
+
+      const p = this.addToResolver_(xhrConsumeErrors)
+          .then(errors => this.verify_(errors))
+          .then(updatedElements => afterVerify(updatedElements));
+
+      if (getMode().test) {
+        this.xhrVerifyPromise_ = p;
+      }
+    }
+  }
+
+  /**
+   * Prevent race conditions from XHRs that arrive out of order by resolving
+   * only the most recently initiated XHR.
+   * TODO(cvializ): Replace this when the Fetch API adds cancelable fetches.
+   * @param {!Promise} promise
+   * @return {!Promise} The resolver result promise
+   */
+  addToResolver_(promise) {
+    if (!this.xhrResolver_) {
+      this.xhrResolver_ = new LastAddedResolver();
+      const cleanup = () => {
+        this.xhrResolver_ = null;
+      };
+      this.xhrResolver_.get().then(cleanup, cleanup);
+    }
+    this.xhrResolver_.add(promise);
+    return this.xhrResolver_.get();
+  }
+
+  /**
+   * Set errors on elements that failed verification, and clear any
+   * verification state for elements that passed verification.
+   * @param {!Array<!VerificationErrorDef>} errors
+   * @private
+   */
+  verify_(errors) {
+    const errorElements = [];
+    let updatedElements = [];
+
+    // Set the error message on each element that caused an error.
+    for (let i = 0; i < errors.length; i++) {
+      const error = errors[i];
+      const element = this.form_./*OK*/querySelector(`[name="${error.name}"]`);
```

Comment:
Neither can match the parent. The issue is `#querySelector` is scoped to the document, then filtered to elements inside the parent (`div span` can match a `<div>` _outside_ parent and then a `<span>` inside parent). `scopedQuerySelector` however, is scoped to the parent only (`div span` can only match a `<div>` inside parent and a `<span>` inside that `<div>`).

- - -

`#querySelector` is allowed when there is a single string without any spaces (meaning, you can not have '* *'). Because this is a template string, it could do all kinds of weird things. But we know better than the linter in this case.